src/main/java/io/jenkins/plugins/checks/github/GitHubChecksPublisherFactory.java

```diff
@@ -37,42 +45,67 @@ public GitHubChecksPublisherFactory() {
-        if (gitSCMContext.isValid(logger)) {
-            return Optional.of(new GitHubChecksPublisher(gitSCMContext, getListener(listener)));
+        try {
+            return createPublisher(run, DisplayURLProvider.get().getRunURL(run), listener);
+        }
+        catch (UnsupportedEncodingException e) {
+            LOGGER.log(Level.WARNING, "Could not create logger.", e);
         }
 
-        return createPublisher(listener, logger, new GitHubSCMSourceChecksContext(run, runURL, scmFacade));
+        return Optional.empty();
     }
 
     @Override
     protected Optional<ChecksPublisher> createPublisher(final Job<?, ?> job, final TaskListener listener) {
-        return createPublisher(job, DisplayURLProvider.get().getJobURL(job), listener);
+        try {
+            return createPublisher(job, DisplayURLProvider.get().getJobURL(job), listener);
+        }
+        catch (UnsupportedEncodingException e) {
+            LOGGER.log(Level.WARNING, "Could not create logger.", e);
+        }
+
+        return Optional.empty();
     }
 
     @VisibleForTesting
-    Optional<ChecksPublisher> createPublisher(final Job<?, ?> job, final String jobURL, final TaskListener listener) {
-        PluginLogger logger = createLogger(getListener(listener));
+    Optional<ChecksPublisher> createPublisher(final Run<?, ?> run, final String runURL, final TaskListener listener)
+            throws UnsupportedEncodingException {
+        ByteArrayOutputStream cause = new ByteArrayOutputStream();
+        try (PrintStream ps = new PrintStream(cause, true, StandardCharsets.UTF_8.name())) {
+            PluginLogger causeLogger = new PluginLogger(ps, "GitHub Checks");
+
+            GitSCMChecksContext gitSCMContext = new GitSCMChecksContext(run, runURL, scmFacade);
+            if (gitSCMContext.isValid(causeLogger)) {
+                return Optional.of(new GitHubChecksPublisher(gitSCMContext, createConsoleLogger(getListener(listener))));
+            }
+
+            GitHubSCMSourceChecksContext gitHubSCMSourceContext = new GitHubSCMSourceChecksContext(run, runURL, scmFacade);
+            if (gitHubSCMSourceContext.isValid(causeLogger)) {
+                return Optional.of(new GitHubChecksPublisher(gitHubSCMSourceContext, createConsoleLogger(getListener(listener))));
+            }
+        }
 
-        return createPublisher(listener, logger, new GitHubSCMSourceChecksContext(job, jobURL, scmFacade));
+        listener.getLogger().print(cause.toString(StandardCharsets.UTF_8.name()));
+        return Optional.empty();
     }
 
-    private Optional<ChecksPublisher> createPublisher(final TaskListener listener, final PluginLogger logger, 
-            final GitHubChecksContext gitHubSCMSourceContext) {
-        if (gitHubSCMSourceContext.isValid(logger)) {
-            return Optional.of(new GitHubChecksPublisher(gitHubSCMSourceContext, getListener(listener)));
+    @VisibleForTesting
```

Comment:
agree with the check above about duplication can it be improved so we don't have 3 very similar methods?